lib/matplotlib/cm.py

```diff
@@ -177,23 +136,67 @@ def register(self, cmap, *, name=None, force=False):
             registered name. True supports overwriting registered colormaps
             other than the builtin colormaps.
         """
+        if not isinstance(cmap, colors.Colormap):
+            raise ValueError("You must pass a Colormap instance. "
+                             f"You passed {cmap} a {type(cmap)} object.")
+
         name = name or cmap.name
-        if name in self and not force:
-            raise ValueError(
-                f'A colormap named "{name}" is already registered.')
-        register_cmap(name, cmap.copy())
+        if name in self:
+            if not force and name in self._builtin_cmaps:
+                msg = f"Trying to re-register the builtin cmap {name!r}."
+                raise ValueError(msg)
+            elif not force:
+                raise ValueError(
+                    f'A colormap named "{name}" is already registered.')
+            else:
+                _api.warn_external(f"Trying to register the cmap {name!r} "
+                                   "which already exists.")
```

Comment:
Would reorder to
```suggestion
            if force:
                _api.warn_external(f"Trying to register the cmap {name!r} "
                                   "which already exists.")
            else:
                raise ValueError(
                    f"Trying to re-register the builtin cmap {name!r}."
                    if name in self._builtin_cmaps else
                    f'A colormap named "{name}" is already registered.')
```

Reply:
Thanks! This feels a bit awkward to me in general... I'm actually wondering if want to outright refuse to override a builtin? Right now we could use `force=True` and replace the builtin, but in unregister we don't allow that removal.

Reply:
Good point. I was just cleaning up the structure, not considering something else. But I agree: We should not allow overriding builtins. Please change.

Reply:
Oops, I forgot to look at our current function. Unfortunately `cm.register_cmap()` doesn't use `force`, it uses `override_builtin`, even more explicitly allowing this behavior. So, I'm not sure how we want to handle this. I really don't think people should be changing a colormap in their underlying library and then re-registering it as the same name, but apparently there are a few (~2) cases that do that :(
https://github.com/search?q=override_builtin&type=issues

Reply:
We have `override_builtin` for compatibility (https://github.com/matplotlib/matplotlib/pull/18503#issuecomment-890600347) and at least for now cannot get rid of it. However, I don't want to have that in the new ColormapRegistry API. Since we're here switching to having the registry as the underlying structure, we have to feed that parameter in privately. Either using a private parameter

```
def register_cmap(...):
    ...
    _colormaps.register(cmap, name=name, force=True, _override_builtin=override_builtin)
```

or using a private flag (with uglier code, but slight peference since it does not show up in the docs):

```
def register_cmap(...):
    ...
    _colormaps._allow_override_builtin = override_builtin
    _colormaps.register(cmap, name=name, force=True)
    _colormaps._allow_override_builtin = False
```